### ISSUE TITLE
Simultaneous segmentation of nucleus and cell

### ISSUE THREAD

**Issue by chrisroat** [MouseLand/cellpose#188]
Is there a way envisioned to predict both nucleus and cell simultaneously?  Currently, the two segmentations can be done independently, though I note that the cell segmentation may have access to the nucleus stain.

The benefits are two-fold:
- The nucleus (or nuclei) of a cell can be linked to the corresponding cell, rather than requiring post-processing to match.
- In cases where the data isn't perfect*, a joint estimation can provide more reasonable estimate.

* Examples:
- Very low RNA expression -- the nucleus would be the best estimate of the cell in that case.
- The nucleus segment is estimated larger than the cell segment for some reason - a constrained estimation might shrink the nucleus and/or enlarge the cell.

**Comment by marius10p** [comment]
I agree it would be nice, but I don't see a direct way to enforce such a constraint. The best we can do is show the neural network a ton of examples where the constraint holds, and hope that it learns to only produce constrained segmentations. 

With that in mind, what you could do is train Cellpose on just the images with nucleus and cytoplasm, and train it to output the flow and probability maps for each. This shouldn't be too difficult to do, but we won't be able to do it ourselves soon. However, the training data and training scripts are public, so anyone who feels up for it could give it a shot.

**Comment by pr4deepr** [comment]
Is it possible to import labelled nuclei??
For example, I have a model trained on segmenting nuclei with high accuracy in StarDist 2D/3D. Can you import this labelled nuclei image into Cellpose and use it for predicting the cell/cytoplasm??

**Comment by carsen-stringer** [comment]
no @pr4deepr we unfortunately don't have a way to use the masks from the nuclei.

We don't have any training data where the nuclei and the cytoplasm are labelled. But once this [data](https://www.biorxiv.org/content/10.1101/2021.03.01.431313v2.full) becomes available we can train a model and I think it will have the benefits you described.

**Comment by pr4deepr** [comment]
Hi
   Thanks for the reply and pointing to this paper. It was an extremely
interesting read!
Also, awesome to see cellpose on Napari..

Looking forward to the future models!!

Cheers
Pradeep

**Comment by adrtsc** [comment]
Hi,
I think it would be extremely helpful to be able to give the nuclear masks as input to the cytoplasm model. When doing nuclear and cytoplasm segmentation of the same cells it often happens that the nuclei are not completely contained in the cytoplasm masks (e.g. some cytoplasm masks contain two full nuclei, or one nuclei and some pixels of another one). For image anaylsis and feature measurements, it is often necessary that there is a 1:1 relationship between objects (e.g. each cytoplasm mask only contains one nuclear mask).

Instead of using training data that contains cells and nuclei to train a new model (that would probably not be perfect and still have cases in which there is no 1:1 relationship), do you think there is some way to use the nuclear masks as a seed to enforce that each cytoplasm mask in the end contains only one nucleus mask and that there is no "spill-out" of nucleus mask pixels out of the corresponding cytoplasm boundaries? 

Thanks!

**Comment by carsen-stringer** [comment]
we were able to achieve state-of-the-art segmentation on the cytoplasm masks from mesmer without using the nuclear masks. for dual segmentation we still recommend segmenting both separately

you could curate your cellular ROIs using your nuclear ROIs, but we don't have plans for implementing that ourselves in cellpose
